### ISSUE TITLE
List::MoreUtils::XS is no longer installed by default when XS is available

### ISSUE THREAD

**Issue by Grinnz** [perl5-utils/List-MoreUtils#25]
This could be considered a regression due to the split dist, as the XS code was always installed before. Unless it's specifically requested as a dependency, List::MoreUtils::XS will not be installed anymore in a fresh install. For reference, List::SomeUtils solves this with dynamic config and ExtUtils::HasCompiler (via dzil plugins): https://metacpan.org/source/DROLSKY/List-SomeUtils-0.53/Makefile.PL#L65. Similarly JSON::MaybeXS using ExtUtils::CBuilder but with much more complicated logic due to JSON::XS issues: https://metacpan.org/source/ETHER/JSON-MaybeXS-1.003009/Makefile.PL#L88. In both of these cases the decision is made based on whether can_xs is true (as defined by the appropriate module) and whether PUREPERL_ONLY is specified.

**Comment by rehsack** [comment]
Since List::MoreUtils::XS needs List::MoreUtils as loader, it is not possible to install it automatically when XS is found. Even more: there is no trivial way to reliable detect XS availability.
So I had to make a decision. I made the decision when reworking issues beginning with 0.400 - and I rethink last 6 month where the code was under development.

A regression in my understanding means, that something what worked before don't work now. But there is nothing what couldn't be done by List::Moreutils::PP - and those who explicitely need XS had always require List::MoreUtils::XS to state so.

**Comment by shadowcat-mst** [comment]
This is a massive performance regression, and for some production loads means an upgrade of List::MoreUtils will result in production services failing.

installing the XS component by default if possible is standard. If you still insist on using your own build tools, you could at least make them work properly.

**Comment by karenetheridge** [comment]
> Since List::MoreUtils::XS needs List::MoreUtils as loader, it is not possible to install it automatically when XS is found.

Then you made the wrong decision in the way these modules depend on each other.

Also, passing `PUREPERL_ONLY=<anything>` causes a build failure, which is unacceptable.  Some cpan clients will pass this value.

`PUREPERL_ONLY=0` should result in the XS module being installed/upgraded.  `PUREPERL_ONLY=1` should always work unless an older version of the XS module is already installed, in which case a failure is reasonable because the user is providing conflicting information and it would be difficult to prevent the older XS bindings from being loaded.

The way you (try to) manually detect conflicts in Makefile.PL is wrong and will result in too many false positives.

I can't imagine how vendor packagers will be able to deal with this Makefile.PL at all.

**Comment by rehsack** [comment]
I think over that and find a solution

**Comment by karenetheridge** [comment]
> I think over that

Come talk to \#toolchain --  many different permutations of PP/XS have been tried, and there is some wisdom to be gained from people who have tried and fallen down in the past :)

**Comment by rehsack** [comment]
Never again

**Comment by shadowcat-mst** [comment]
Opened a #26 to give us a fresh discussion of the specific issues.

You should try #toolchain again eventually though - it's been far more civilised since riba's tantrum+ragequit.

**Comment by Leont** [comment]
> Since List::MoreUtils::XS needs List::MoreUtils as loader, it is not possible to install it automatically when XS is found.

Could you explain why this is necessary? 

>  Even more: there is no trivial way to reliable detect XS availability.

I am surprised by this statement coming from you.

> A regression in my understanding means, that something what worked before don't work now. But there is nothing what couldn't be done by List::Moreutils::PP - and those who explicitely need XS had always require List::MoreUtils::XS to state so.

I may be mistaken, but I had the impression that they're not identical in the way they're dealing with `@_`.

**Comment by rehsack** [comment]
@shadowcat-mst - sorry when I do not believe, even you're engagement speaks another language.

> > Since List::MoreUtils::XS needs List::MoreUtils as loader, it is not possible to install it automatically when XS is found.
>
> Could you explain why this is necessary?

@Leont - it's just developed that way because when designing pure LMU at niederrhein.pm ages ago, this was simply recommended by all discussion attendees.
You can call it a mistake or bad design or whatever - it's at the moment at it is and was out for nearly 8 weeks and could be reviewed.

> > Even more: there is no trivial way to reliable detect XS availability.
>
> I am surprised by this statement coming from you.

Well, the split was done to remove C::AC from List::MoreUtils ...
Clearer?

> > A regression in my understanding means, that something what worked before don't work now. But there is nothing what couldn't be done by List::Moreutils::PP - and those who explicitely need XS had always require List::MoreUtils::XS to state so.
> 
> I may be mistaken, but I had the impression that they're not identical in the way they're dealing with @_.

That would be bad. Please file tickets when you find such a behavior when it should be identical.

**Comment by shadowcat-mst** [comment]
@rehsack just closed #26 as well. We need to either re-open one of these issues or somebody needs to file a new one that @rehsack promises to listen to instead of closing them every time somebody points out he made a mistake.

**Comment by rehsack** [comment]
@shadowcat-mst - I promise to listen to any reasonable point. But I do not listen to shout contest.

**Comment by shadowcat-mst** [comment]
@rehsack and yet in spite of saying that asking to re-open an issue is fine, in spite of there being a valid problem here, you've left both issues closed.

Please pick one to re-open or create a new one or give us permission to create a new one without you claiming it's "pressuring".

Leaving both closed while telling me off for creating a still open one is not constructive.

**Comment by rehsack** [comment]
Oh yes, if you like it being open, I'm doing that for you. I don't mind - since I realized the point and don't need the open issue to remember.

**Comment by shadowcat-mst** [comment]
To summarise where we're currently up to:

1) List::MoreUtils must install ::XS by default if a compiler is available

2) PUREPERL_ONLY=1 must override this and ensure List::MoreUtils is installed pp-only in that case

3) PUREPERL_ONLY=0 must be a no-op rather than crashing Makefile.PL

Provided all three of these are true, exactly how you ensure they are true is up to you.

These three requirements reflect the standard requirements of PP/XS splits, as previously discussed when we were comparing the standard build approaches for low level modules to your new/preferred approaches.

**Comment by kentfredric** [comment]
> I can't imagine how vendor packagers will be able to deal with this Makefile.PL at all.

As a vendor, I'm just going to say that this specific interaction is likely not going to be a problem.

Mostly, because if I want to make users have the ability to install ::XS, I can do just that without needing to have Makefile.PL make it happen.

The only fun part for me is making sure the Makefile.PL doesn't spontaneously `die()` on me in either condition and respects the wishes the user indicated.

Which means the only really hard thing is understanding the flow of the code in Makefile.PL, and identifying how to push it in the right direction, and if the code doesn't allow this as-is, understanding the code sufficient so I can patch it to do the right thing.

*Ideally* though, we do encourage upstream to make their Makefile.PL's such that we can authoritatively declare what we want out of it, and have it respect that, as that is typically superior to maintaining our own patches ( because the further we deviate from upstream, the more work there is to continue maintaining upstreams releases )

Granted: I haven't yet assessed the nature of the latest LMU updates, and there might be problems in there other than this magic. I'm just not jumping the gun until I find a problem that is too hard to circumvent downstream.

In short though: we can make LMU do the right thing, even if its too hard to do the right thing on CPAN.
"configuration avoiding automagic" is the only thing that makes me truely miserable.

**Comment by rehsack** [comment]
> As a vendor, I'm just going to say that this specific interaction is likely not going to be a problem.
> 
> Mostly, because if I want to make users have the ability to install ::XS, I can do just that without needing to have Makefile.PL make it happen.

I expect that @shadowcat-mst see's his opan as a vendor-tool, either. I don't want to discuss that here (separation of concerns) and for the moment I accept this as a sane reason.

> The only fun part for me is making sure the Makefile.PL doesn't spontaneously die() on me in either condition and respects the wishes the user indicated.

If there is something nasty on Makefile.PL's behavior (in doubt, don't try to understand the code - just tell me what you expect and what you got and maybe why you think that's wrong), I'll fix that immediately (see https://rt.cpan.org/Ticket/Display.html?id=120799). This thread is about mislead design decisions and is somewhere between serious issue and wishlist. Kind of both - so (as long as this thread doesn't start eating time by spamming comments (not yours), it'll be in line directly after the blead-perl fixes (blockers) of some other distributions of me).

> Which means the only really hard thing is understanding the flow of the code in Makefile.PL, and identifying how to push it in the right direction, and if the code doesn't allow this as-is, understanding the code sufficient so I can patch it to do the right thing.

This shouldn't be necessary.

> Ideally though, we do encourage upstream to make their Makefile.PL's such that we can authoritatively declare what we want out of it, and have it respect that, as that is typically superior to maintaining our own patches ( because the further we deviate from upstream, the more work there is to continue maintaining upstreams releases )

And I will gracefully support this approach as long as it doesn't try to add insane code snippets ;)

> Granted: I haven't yet assessed the nature of the latest LMU updates, and there might be problems in there other than this magic. I'm just not jumping the gun until I find a problem that is too hard to circumvent downstream.

> In short though: we can make LMU do the right thing, even if its too hard to do the right thing on CPAN.
> "configuration avoiding automagic" is the only thing that makes me truely miserable.

I can't parse the last sentence. Is there something you expect from Makefile.PL to do or to avoid?

**Comment by rehsack** [comment]
> To summarise where we're currently up to:

> 1) List::MoreUtils must install ::XS by default if a compiler is available

The primary reason to split PP and XS was to remove the XS-availability stuff. I will not re-introduce them.

> 2) PUREPERL_ONLY=1 must override this and ensure List::MoreUtils is installed pp-only in that case

PUREPERL_ONLY=1 will not install ::XS

> 3) PUREPERL_ONLY=0 must be a no-op rather than crashing Makefile.PL

PUREPERL_ONLY=0 can recommend or suggest the ::XS.
It'll not test anything.

> 4) List::MoreUtils::XS must function on its own so we don't have circular dep problems with (1)

Agreed.

> Provided all three of these are true, exactly how you ensure they are true is up to you.

Four and no. (1) is not reasonable compared to former requirements (se #9 and #11). If this discussion explodes again because of this topic - we close and don't come back. (3) bases on (1) and is therefore not applicable in the described way.

> These requirements reflect the standard requirements of PP/XS splits, as previously discussed when we were comparing the standard build approaches for low level modules to your new/preferred approaches.

IEEE nnnn? ANSI ??? - don't throw with "standard" phrases as standards are the last thing you really care. Standards are bigger than Perl(5) - and we don't want to discuss "standards" here (otherwise explode, bla bla).

**Comment by kentfredric** [comment]
> I can't parse the last sentence. Is there something you expect from Makefile.PL to do or to avoid?

At this point I don't have any specific requests, I'm mostly just trying to give a more objective view of what vendor concerns look like, in contrast to ether's suggestion they'll be too difficult.

If I have anything specific I need fixed, I'll work that out next week sometime when I get around to packaging the latest release. :) 

But most of my commentary so far in this thread is not necessarily just for you, but for the other participants, in effort to give a better understanding of the general needs of downstream.

Because while downstream does have a different set of needs than CPAN, they also have a different set of powers, and we can pull fancy tricks that aren't possible on CPAN :)

**Comment by haarg** [comment]
Adding List::MoreUtils::XS to suggests or recommends dynamically isn't going to work well, unfortunately.  Adding suggests or recommends would be done with META_ADD or META_MERGE, since EUMM doesn't support a parameter to specify them.  But META_ADD and META_MERGE are ignored when configuring on user machines.  You would need a hack of some kind to get EUMM to adjust suggests or recommends dynamically.

**Comment by rehsack** [comment]
Unfortunately you (requestors) cannot have all - No C::AC, XS depend on PP, XS standalone, bla bla bla - it is insane.

What can be done is adding it (hard!) to requirements and remove it from there dynamically when PUREPERL_ONLY (or similar) is set appropriately.

**Comment by rehsack** [comment]
This issue is solved. Further details are discussed in https://rt.cpan.org/Ticket/Display.html?id=122875

**Comment by Grinnz** [comment]
I don't agree that this issue is solved. Recommends is not a solution, as no CPAN installers use it by default. Relying only on PUREPERL_ONLY=1 for a hard dependency would be acceptable but unfortunate, detecting the lack of a compiler automatically and using PUREPERL_ONLY=1 only as a manual override is much more user-friendly.

**Comment by Grinnz** [comment]
This seems to be fixed now with 0.425.
